api/webhook/github.ts

```diff
@@ -420,8 +330,101 @@ async function initiateBuild(
     });
   }
 
+  const entry = await database.getModule(moduleName);
+
+  const resp = await checkModuleInfo(
+    entry,
+    moduleName,
+    owner,
+    sender,
+    repo_id,
+    subdir,
+  );
+  if (resp) return resp;
+
+  // Update meta information in MongoDB (registers module if not present yet)
+  await database.saveModule({
+    ...entry ??
+      {
+        name: moduleName,
+        type: "github",
+        created_at: new Date(),
+        is_unlisted: false,
+      },
+    repo_id,
+    owner,
+    repo,
+    description,
+    star_count: starCount,
+  });
+
   const version = ref.substring(versionPrefix.length);
+  const invalidVersion = await checkVersion(moduleName, version);
+  if (invalidVersion) return invalidVersion;
+
+  const buildID = await database.createBuild({
+    options: {
+      type: "github",
+      moduleName,
+      repository: `${owner}/${repo}`,
+      ref,
+      version,
+      subdir: subdir ?? undefined,
+    },
+    status: "queued",
+  });
 
+  await queueBuild(buildID);
+
+  return respondJSON({
+    statusCode: 200,
+    body: JSON.stringify({
+      success: true,
+      data: {
+        module: moduleName,
+        version,
+        repository: `${owner}/${repo}`,
+        status_url: `https://deno.land/status/${buildID}`,
+      },
+    }),
+  });
+}
+
+/**
+ * CheckModuleInfo performs a series of general validation on the module such as
+ * validating that the module name complies with the naming convention enforced
+ * on deno.land/x, whether or not the sender or owner have been blocked etc...
+ *
+ * These verifications are meant to be performed before the module is registered
+ * or updated in the database to prevent "bad" modules from being pushed to the
+ * build queue and subsequently published.
+ *
+ * @param entry database entry for the module
+ * @param moduleName module name as shown on deno.land/x
+ * @param owner username of the GH repository owner
+ * @param sender username of the user triggering the webhoo
+ * @param repo_id numerical id of the GH repository
+ */
+async function checkModuleInfo(
+  entry: Module | null,
+  moduleName: string,
+  owner: string,
+  sender: string,
+  repo_id: number,
+  subdir: string | null,
+): Promise<APIGatewayProxyResultV2 | undefined> {
+  return checkSubdir(subdir) ??
+    await checkBlocked(sender) ??
+    await checkBlocked(owner) ??
+    checkMatchesRepo(entry, repo_id) ??
+    await checkModulesInRepo(repo_id) ??
+    await hasReachedQuota(owner) ??
+    await checkName(moduleName);
```

Comment:
These checks should not happen when the module is already in the database (to support grandfathered bad names). Currently this is a functional change.

Reply:
do we also have grandfathered modules that need to be exempt from the quota and modules-per-repo rules as well?

Reply:
Well I don't want modules owners to be locked out from publishing, because they are at limit. So yes, effectively once you have published the first version for a module, it is exempt from all quota checks except blocking. The quota checks should only apply to new modules.

Reply:
I'll add test cases for those, and I'll do the same as I did for the repo_id check and push the check for `!entry` down to each of those three functions to keep this one as straight forward as possible.